### ISSUE TITLE
Cleanup warning when logging out from Zabbix API

### ISSUE THREAD

**Issue by dimir** [whosgonna/Zabbix-Tiny.pm#26]
When script exits the following warning (not affecting the exit code) appears:

>         (in cleanup) Can't call method "post" on an undefined value at /usr/local/share/perl/5.24.1/Zabbix/Tiny.pm line 200 during global destruction.

The issue happens in fuction DEMOLISH on issuing a post to log out, I guess the user agent object is destroyed at this point that's why the error:

`$self->{post_response} = $ua->post( $url, @content_type, Content => $json );`

I fixed it by applying a very simple patch:

```
$ diff -u /usr/local/share/perl/5.24.1/Zabbix/Tiny.pm.bak /usr/local/share/perl/5.24.1/Zabbix/Tiny.pm
--- /usr/local/share/perl/5.24.1/Zabbix/Tiny.pm.bak     2018-02-13 14:53:03.000000000 +0530
+++ /usr/local/share/perl/5.24.1/Zabbix/Tiny.pm 2018-02-13 14:53:14.000000000 +0530
@@ -187,6 +187,9 @@
     my $ua        = $self->ua;
     my $auth      = $self->auth;
     my $url       = $self->server;
+
+    return unless ($ua);
+
     my $json_data = {
         jsonrpc => '2.0',
         id      => ++$self->{ id },

```

Using Tiny 1.09 on Debian stable, perl version 5.24.1 .

**Comment by whosgonna** [comment]
Thank you. This seems pretty straightforward and sane.  Do you plan to submit a pull request, or just want me to do this?    I'm guessing that the problem can be recreated by instantiating the Zabbix::Tiny object, and then exiting without ever creating a connection?  I can write a test for this as well.

**Comment by dimir** [comment]
Sorry for a late response.

If you could do the patching that'd be great. :-)

I think I was getting this error after successfully performing operations through Zabbix API though... I'll try to get back to this soon.

**Comment by dimir** [comment]
I just checked that one again. Actually the error appears after instantiating and performing some API calls. So the object is instantiated and used successfully before the warning appears. I have 2 identical boxes running Red Hat Enterprise Linux Server release 6.6 (Santiago). Some perl modules are installed as rpm packages, some through CPAN, different combinations on the boxes. So on one box I get the error, on another - not. Both Zabbix::Tiny are installed through CPAN, version 1.07. They are identical. So the difference must be somewhere in dependencies. E. g. on the box where the error is reproduced Capture::Tiny is version 0.46 while where it is not - 0.24 .

I have Zabbx::Tiny on one newer box running Red Hat Enterprise Linux Server release 7.4 (Maipo) where the issue is not reproducible.

If you have any ideas where I could look further I'd be thankful.

P. S. Just remembered that it is also reproducible on my Debian box running stable (9.3).

P. P. S. This could be related as I suspect what happens is something like DEMOLISH is called 2 times. The module Moo.pm where this is reproducible is 2.003000, while where not - 1.004002.

**Comment by whosgonna** [comment]
Thank you for the updates.  I think anyway it goes, checking that the LWP::UserAgent object exists prior to calling log-out is a good idea.    I still haven't had a chance to reproduce, and I'd like to have a solid way to make the problem happen, so that I can add a test case that proves the fix into the future.  Work in progress.

**Comment by whosgonna** [comment]
Can you also check your version of LWP?  I'm running most of my testing on FreeBSD, but Perl 5.24.3, Moo 2.003004, so that's quite similar.

**Comment by dimir** [comment]
Where it's reproducible:

```
$ ./print-module-version.pl LWP LWP::UserAgent Moo
LWP                 : 6.06
LWP::UserAgent      : 6.06
Moo                 : 2.003
$ perl --version
This is perl, v5.10.1 (*) built for x86_64-linux-thread-multi
```

Where it's not:

```
$ ./print-module-version.pl LWP LWP::UserAgent Moo
LWP                 : 6.06
LWP::UserAgent      : 6.06
Moo                 : 1.004002

$ perl --version
This is perl, v5.10.1 (*) built for x86_64-linux-thread-multi
```

I have this script to print used module versions, feel free to provide me the bigger list to get more information.

**Comment by whosgonna** [comment]
Fixed in 6155474.

I created a test by manually setting ```ua``` to undef, then explicitly calling ```DEMOLISH``` and making sure that it returns ```undef``` rather than an exception.

I'll take a peek at #27 to improve the documentation before pushing to CPAN.

**Comment by whosgonna** [comment]
Fixed in 4032117.  I'm still not sure what causes this issue, however it's a simple logical test to see if the LWP::UserAgent exists prior to issuing the logout command.  

v1.10 is pushed to CPAN.
